network/p2p/ping.go

```diff
@@ -0,0 +1,174 @@
+package p2p
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-msgio"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/cmd/bootstrap/build"
+	"github.com/onflow/flow-go/network/message"
+)
+
+// the Flow Ping protocol prefix
+const FlowLibP2PPingPrefix = "/flow/ping/"
+
+const pingTimeout = time.Second * 60
+
+// PingService handles the outbound and inbound ping requests and response
+type PingService struct {
+	host           host.Host
+	logger         zerolog.Logger
+	pingProtocolID protocol.ID
+}
+
+func NewPingService(h host.Host, pingProtocolID protocol.ID, logger zerolog.Logger) *PingService {
+	ps := &PingService{host: h, logger: logger, pingProtocolID: pingProtocolID}
+	h.SetStreamHandler(pingProtocolID, ps.PingHandler)
+	return ps
+}
```

Comment:
Suggestion for a different approach than `SetSemver`:

```suggestion
type PingService struct {
	host           host.Host
	logger         zerolog.Logger
	getVersion func() string
	pingProtocolID protocol.ID
}

func NewPingService(h host.Host, pingProtocolID protocol.ID, logger zerolog.Logger, getVersion func() string) *PingService {
	ps := &PingService{host: h, logger: logger, pingProtocolID: pingProtocolID, getVersion: getVersion}
	h.SetStreamHandler(pingProtocolID, ps.PingHandler)
	return ps
}
```

Then we can do:
```go
pingService := NewPingService(..., build.Semver)
```

And for testing something like:
```go
pingService := NewPingService(..., func() {return "test string"})
```

Using the function argument directly is simplest and probably fine for now, but if we're adding block height and potentially other things as well, it may be cleaner to create an interface for `PingInfoProvider`
```go
type interface PingInfoProvider {
      SoftwareVersion() string
      LatestFinalizedBlockHeight() uint64
}
```

Reply:
makes sense. for now I am just passing in the semver value as a string to the `NewPingService` function since the software version won't change till the node is restarted.
I will define the interface once I start implementing the finalized block height changes.